Add message for unsupported features

pitlv2109 opened the pull request:
* Add a message for unsupported features
* Add a message to prevent users from blindly applying the converted policy.

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 307aac30024f029f475ecefdfeb31ef12999b325 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18860/lint_istio/3824) | `/test lint_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

istio-testing commented:
@pitlv2109: The following tests **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
lint_istio | 307aac30024f029f475ecefdfeb31ef12999b325 | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18860/lint_istio/3824) | `/test lint_istio`
pilot-e2e-envoyv2-v1alpha3_istio | cb1af74b48720f378be6678235e6f0fe22900bef | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18860/pilot-e2e-envoyv2-v1alpha3_istio/3793) | `/test pilot-e2e-envoyv2-v1alpha3_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
/test pilot-e2e-envoyv2-v1alpha3_istio

howardjohn commented:
this seems like a really bad UX. If a user wants to review it before applying they can just not pipe to kubectl. Adding arbitrary hoops for users to jump through would be really frustrating

pitlv2109 commented:
> ... they can just not pipe to kubectl

I think this is something I don't want to bet, especially given authorization policies are important to the system and misconfiguring them can have significant impact. This is to prevent someone from running `convert command | kubectl apply -f -` without realizing it.

howardjohn commented:
@pitlv2109 the same applies to Helm install (actually this is arguable even more important than auth rules), but we don't have users complain they accidentally `kubectl apply`ed the output of that. If a tool added this sort of output I would think very hard about whether I wanted to us it. Its a terrible experience for users

yangminzhu commented:
Another question, the PR description mentions `Fix default trust domain` but I'm not sure which part of this PR is doing that? @pitlv2109

yangminzhu commented on istioctl/cmd/testdata/authz/converter/service-role-with-constraints.yaml:
```diff
@@ -0,0 +1,12 @@
+apiVersion: "rbac.istio.io/v1alpha1"
```
When testing this file manually, I found the following generated in stdout in addition to the error message in stderr:
```
map[destination.ip:[10.1.2.3 10.2.0.0/16]]
```
We should not generate anything to stdout when an error happens, and ideally the above message should be generated to stderr as well.

pitlv2109 replied:
I have a print statement that I forgot to remove :)

pitlv2109 commented:
> Another question, the PR description mentions `Fix default trust domain` but I'm not sure which part of this PR is doing that? @pitlv2109

I added some code to fix the empty trust domain part just in case the other PR didn't get in in time. Removed the description now. Thanks

yangminzhu commented:
> this seems like a really bad UX. If a user wants to review it before applying they can just not pipe to kubectl. Adding arbitrary hoops for users to jump through would be really frustrating

@pitlv2109 can we also update the help text of the `authz` sub-command to mention that user needs to review and remove the line `===PLEASE REVIEW THE GENERATED POLICY AND REMOVE THIS LINE BEFORE APPLYING IT===` in the generated policy?

@howardjohn Maybe we could add a new flag, e.g. `--no-confirm`, The line is not generated when the flag is used, the point here is to force the user to double check the generated files as the tool is still considered experimental.

howardjohn commented:
> @howardjohn Maybe we could add a new flag, e.g. --no-confirm, The line is not generated when the flag is used, the point here is to force the user to double check the generated files as the tool is still considered experimental.

Why not just add a doc like

> Run `istioctl convert` and check the output. Look for x y z. Once you are satisified, run `istioctl convert | kubectl apply`

Or something?

pitlv2109 commented:
The confirmation line is just one line and is always on top of the converted policy. I don't think it's that hard to just remove one line. I'd rather make it safer than betting on users reading docs

howardjohn commented:
> The confirmation line is just one line and is always on top of the converted policy. I don't think it's that hard to just remove one line. I'd rather make it safer than betting on users reading docs

Normal UX:
Run command once as dry run, run again to commit. Everyone understands this because this is how 99% of tools work

Proposed UX:
Run command once as dry run, then run again to commit. Get obscure error message from kubectl. Look again and realize we decided to put a bunch of junk in the file. Rerun the command a redirect to a file. Open up that file in an editor, and go through and the junk, save. Run `kubectl file`. 

Its just a lot of churn for no reason. I have never ever seen a tool do this before.

pitlv2109 commented:
@yangminzhu Updated the usage. Could you take a look?

pitlv2109 commented:
> Run command once as dry run

This is again betting the users will read docs.

> Get obscure error message from kubectl. Look again and realize we decided to put a bunch of junk in the file.

 What do you mean a bunch of junk? The only thing it does it prepend one string on top of the whole converted policies. The other are straight error messages about semantics that the tool does not support yet, and it won't generate any converted policies.

esnible commented:
In `manifest apply` we use `--skip-confirmation`.

howardjohn commented:
> In manifest apply we use --skip-confirmation.

Its different though because in one case you type `y` and confirm and apply it to your cluster. In this case you have to do manual text editing

yangminzhu commented:
> > In manifest apply we use --skip-confirmation.
> 
> Its different though because in one case you type `y` and confirm and apply it to your cluster. In this case you have to do manual text editing

@esnible @howardjohn The intention is to **force** users to really take a look at the converted policies before applying them. It would be very easy to remove the line in the beginning of the file and we're concerned that some users may run the tool and blindly apply the converted policies to their cluster which may cause vulnerabilities if the tool is somehow not working correctly.

The tool is supposed to be used as a starting point of converting the `v1alpha1` policy instead of letting the tool alone to complete all the work.

Beyond that, I'm okay with other parts of the PR. cc @pitlv2109

pitlv2109 commented:
@esnible lmk if you have any concern about this PR

esnible commented on pilot/pkg/security/authz/converter/converter.go:
```diff
@@ -427,6 +432,9 @@ func convertAccessRuleToOperation(rule *authz_model.Permission) (*rbac_v1beta1.O
 		return nil, fmt.Errorf("invalid input: No rule found in ServiceRole")
 	}
+	if len(rule.Constraints) > 0 && len(rule.Constraints[0]) > 0 {
+		return nil, fmt.Errorf("serviceRole with constraints is not supported")
```
Nit: ServiceRole should be capitalized.  The usual Go Rule doesn't apply [ https://github.com/golang/go/wiki/Errors#errors ] because ServiceRole is a proper name.  Use `// nolint: gas` to get around Lint.

pitlv2109 replied:
Done. PTAL

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
e2e-simpleTests-distroless_istio | fe13ca6f67d5fbb40f6184dc8487fd7d3041041a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18860/e2e-simpleTests-distroless_istio/3933) | `/test e2e-simpleTests-distroless_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
/test e2e-simpleTests-distroless_istio

pitlv2109 commented:
@esnible ping

istio-testing commented:
@pitlv2109: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
e2e-simpleTestsMinProfile_istio | fe13ca6f67d5fbb40f6184dc8487fd7d3041041a | [link](https://prow.istio.io/view/gcs/istio-prow/pr-logs/pull/istio_istio/18860/e2e-simpleTestsMinProfile_istio/3934) | `/test e2e-simpleTestsMinProfile_istio`



<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>
<!-- test report -->

pitlv2109 commented:
/test e2e-simpleTestsMinProfile_istio

istio-testing commented:
In response to a cherrypick label: new pull request created: #19002
